Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -0,0 +1,303 @@
+                    return false;
+                }
+
+                return true;
+            //We don't want to place a space tile on something that's already a space tile. Let's do the inverse of the last check.
+            case RcdMode.Deconstruct:
+                if (tile.Tile.IsEmpty)
+                {
+                    return false;
+                }
+
+                //They tried to decon a turf but the turf is blocked
+                if (target == null && tile.IsBlockedTurf(true))
+                {
+                    ClientPopup(Loc.GetString("rcd-component-tile-obstructed-message"), uid, user);
+                    return false;
+                }
+                //They tried to decon a non-turf but it's not in the whitelist
+                if (target != null && !_tag.HasTag(target.Value, "RCDDeconstructWhitelist"))
+                {
+                    ClientPopup(Loc.GetString("rcd-component-deconstruct-target-not-on-whitelist-message"), uid, user);
+                    return false;
+                }
+
+                return true;
+            //Walls are a special behaviour, and require us to build a new object with a transform rather than setting a grid tile, thus we early return to avoid the tile set code.
+            case RcdMode.Walls:
+                if (tile.Tile.IsEmpty)
+                {
+                    ClientPopup(Loc.GetString("rcd-component-cannot-build-wall-tile-not-empty-message"), uid, user);
+                    return false;
+                }
+
+                if (tile.IsBlockedTurf(true))
+                {
+                    ClientPopup(Loc.GetString("rcd-component-tile-obstructed-message"), uid, user);
+                    return false;
+                }
+                return true;
+            case RcdMode.Airlock:
+                if (tile.Tile.IsEmpty)
+                {
+                    ClientPopup(Loc.GetString("rcd-component-cannot-build-airlock-tile-not-empty-message"), uid, user);
+                    return false;
+                }
+                if (tile.IsBlockedTurf(true))
+                {
+                    ClientPopup(Loc.GetString("rcd-component-tile-obstructed-message"), uid, user);
+                    return false;
+                }
+                return true;
+            default:
+                return false; //I don't know why this would happen, but sure I guess. Get out of here invalid state!
+        }
+    }
+
+    private void NextMode(EntityUid uid, RCDComponent comp, EntityUid user)
+    {
+        _audio.PlayPredicted(comp.SwapModeSound, uid, user);
+
+        var mode = (int) comp.Mode;
+        mode = ++mode % RCDModeCount;
+        comp.Mode = (RcdMode) mode;
+        Dirty(comp);
+
+        var msg = Loc.GetString("rcd-component-change-mode", ("mode", comp.Mode.ToString()));
+        ClientPopup(msg, uid, user);
+    }
+
+    // TODO: see if this isnt implemented in some utility it seems fairly common
```

Comment:
It's not yet, it just needs something similar to audiosystem.

Reply:
ill put a virtual thing on shared popup system